plugins/experimental/ssl_session_reuse/src/publish.cc

```diff
@@ -0,0 +1,401 @@
+        TSDebug(PLUGIN, "RedisPublisher::setup_connection. Successfully AUTH redis server.");
+        freeReplyObject(reply);
+      }
+      break;
+    }
+
+    TSError("RedisPublisher::setup_connection.connect failed.will wait for: %d microseconds and try again.", m_redisRetryDelay);
+    ::usleep(m_redisRetryDelay);
+  }
+
+  return my_context;
+}
+
+void
+RedisPublisher::teardown_connection(::redisContext *context)
+{
+  if (context) {
+    ::redisFree(context);
+  }
+}
+
+::redisReply *
+RedisPublisher::send_publish(::redisContext **ctx, const RedisEndpoint &re, const Message &msg)
+{
+  ::pthread_t my_id = ::pthread_self();
+  TSDebug(PLUGIN, "RedisPublisher::send_publish.called by threadId: %d", static_cast<int>(my_id));
+
+  ::redisReply *current_reply(NULL);
+
+  for (int i = 0; i < static_cast<int>(m_redisPublishTries); ++i) {
+    if (!(*ctx)) {
+      (*ctx) = setup_connection(re);
+
+      if (!(*ctx)) {
+        TSError("RedisPublisher::send_publish.Unable to setup a connection to the redis server: %s:%d .ThreadId: %d Try: %d",
+                re.m_hostname.c_str(), re.m_port, static_cast<int>(my_id), (i + 1));
+        continue;
+      }
+    }
+
+    current_reply = static_cast<redisReply *>(::redisCommand((*ctx), "PUBLISH %s %s", msg.channel.c_str(), msg.data.c_str()));
+    if (!current_reply) {
+      TSError("RedisPublisher::send_publish.Unable to get a reply from the server for publish.ThreadId: %d Try: %d",
+              static_cast<int>(my_id), (i + 1));
+      teardown_connection(*ctx);
+      *ctx = NULL;
+
+    } else if (REDIS_REPLY_ERROR == current_reply->type) {
+      TSError("RedisPublisher::send_publish.Server responded with error for publish.ThreadId: %d Try: %d", static_cast<int>(my_id),
+              i + 1);
+      clear_reply(current_reply);
+      current_reply = NULL;
+      teardown_connection(*ctx);
+      *ctx = NULL;
+    } else {
+      break;
+    }
+  }
+
+  return current_reply;
+}
+
+void
+RedisPublisher::clear_reply(::redisReply *reply)
+{
+  if (reply) {
+    ::freeReplyObject(reply);
+  }
+}
+
+void
+RedisPublisher::runWorker()
+{
+  ::pthread_mutex_lock(&m_endpointIndexMutex);
+  RedisEndpoint &my_endpoint(m_redisEndpoints[m_endpointIndex]);
+
+  if (static_cast<int>(m_redisEndpoints.size() - 1) == m_endpointIndex) {
+    m_endpointIndex = 0;
+  } else {
+    ++m_endpointIndex;
+  }
+  ::pthread_mutex_unlock(&m_endpointIndexMutex);
+
+  ::pthread_t my_id = ::pthread_self();
+  ::redisContext *my_context(setup_connection(my_endpoint));
+  ::redisReply *current_reply(NULL);
+
+  while (true) {
+    ::sem_post(&m_workerSem);
+    int readyWorkers = 0;
+    ::sem_getvalue(&m_workerSem, &readyWorkers);
+
+    // LOGDEBUG(PUB, "RedisPublisher::runWorker.Ready worker count: " << readyWorkers);
+
+    ::pthread_mutex_lock(&m_messageQueueMutex);
+
+    if (m_messageQueue.empty()) {
+      ::sem_wait(&m_workerSem);
+      ::pthread_mutex_unlock(&m_messageQueueMutex);
+      usleep(1000);
```

Comment:
Using std::deque (with queue entries that may be somewhat expensive to copy, involving dynamic allocation) and polling for entries to dequeue at fixed intervals may be too brute force.  Consider something like https://github.com/ywkaras/MiscRepo/tree/master/IP_QUE instead.